File: utility-benchmarks/python/reduction_ops.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf
import utilities as util
import time

config = tf.ConfigProto()
config.gpu_options.allow_growth = True


def get_reduction_op(x, option):
    reductions = {
        'argMax': tf.argmax(x),
        'argMin': tf.argmin(x),
        'sum': tf.reduce_sum(x),
        'mean': tf.reduce_mean(x),
        'logSumExp': tf.reduce_logsumexp(x)
    }

    assert option in reductions, 'Reduction option not found: ' + option
    return reductions[option]


def run(size, option, backend):
    if backend == 'gpu' and util.can_use_gpu():
        doWarmup = True
    else:
        assert backend == 'cpu'
        doWarmup = False

    x = tf.random_uniform([size, size], -1, 1)
    reduction = get_reduction_op(x, option)

    backend = util.get_backend(backend)
    with tf.device(backend):
        sess = tf.Session()

        if doWarmup:
            sess.run(reduction)

        start = time.time()
        sess.run(reduction)
        end = time.time()
        runtime = end - start
        print(runtime)


if __name__ == "__main__":
    run(100, 'argMax', 'cpu')

File: utility-benchmarks/javascript/reduction_ops.ts
import * as dl from 'deeplearn';

import {BenchmarkTest} from './benchmark';
import * as benchmark_util from './utilities';

function getReductionOp(option: string): (x: dl.Tensor) => dl.Scalar {
    switch (option) {
        case 'argMax':
            return x => x.argMax();
        case 'argMin':
            return x => x.argMin();
        case 'sum':
            return x => x.sum();
        case 'logSumExp':
            return x => x.logSumExp();
        case 'mean':
            return x => x.mean();
        default:
            throw new Error(`Not found such ops: ${option}`);
    }
}

export class ReductionOpsCPU implements BenchmarkTest {
    async run(size: number, option: string): Promise<number> {
        dl.setBackend('cpu');

        const input: dl.Tensor2D = dl.randomUniform([size, size], -1, 1);
        const op = getReductionOp(option);
        const start = performance.now();

        op(input).get();

        const end = performance.now();
        return end - start;
    }
}

export class ReductionOpsGPU implements BenchmarkTest {
    async run(size: number, option: string) {
        dl.setBackend('webgl');

        const input: dl.Tensor2D = dl.randomUniform([size, size], -1, 1);
        const op = getReductionOp(option);

        const benchmark = () => op(input);

        const time = await benchmark_util.warmupAndBenchmarkGPU(benchmark);

        input.dispose();

        return time;
    }
}

export async function run(size, option, backend){
    let benchmark: BenchmarkTest = (backend === 'gpu') ? new ReductionOpsGPU() : new ReductionOpsCPU();
    return await benchmark.run(size, option)
}
